pennylane/measurements/probs.py

```diff
@@ -241,6 +241,43 @@ def process_state(self, state: Sequence[complex], wire_order: Wires):
         # flatten and return probabilities
         return qml.math.reshape(prob, flat_shape)
 
+    def process_counts(self, counts: dict, wire_order: Wires) -> np.ndarray:
+        # checks that 'counts' is a dictionary matching the format returned by 'CountsMP'
+        assert isinstance(counts, dict), "Sampling information should be a dictionary"
+        assert counts, "Sampling information should not be empty"
+
+        for outcome, occurrence in counts.items():
+            assert isinstance(outcome, str), f"Key {outcome} should be a string"
+            assert all(bit in "01" for bit in outcome), f"Key {outcome} should be binary"
+            assert isinstance(occurrence, int), f"Value {occurrence} should be an integer"
+            assert occurrence > 0, f"Value {occurrence} should be positive"
+
+        assert len(set(len(key) for key in counts.keys())) == 1, "Keys should have the same length"
```

Comment:
Our validation strategy for PennyLane tends to be:

1. If it's a mistake that a user is likely to make, or someone did make and got confused by
2. If it's a mistake that is particularly hard to trace back and understand what happened

Otherwise, we take the more pythonic route of "let people do what they want and live with the consequences if they're breaking the requirements". This doesn't bog the performance down with a bunch of validation that adds up over time. While assert statements can be optimized away with `-O`, we can't really assume people are doing that most of the time.

But as always, this is an ongoing discussion that morphs over time.

Long story short, let's assume that the counts dictionary looks like it's supposed to.

Reply:
To add some more context, `MeasurementProcess.process_xyz` are generally developer facing, so it is often okay to leave out such validation in favour of pre-conditions imposed on the function arguments. This is obviously not a good philosophy for user-facing code, but as I said, this is developer-facing (or power-user facing), which allows us to loosen the constraints on the arguments.

Reply:
Thank you both for clarifying this! 

It makes a lot of sense and explains why there are no similar checks in other methods. I'll remove all checks and asserts as soon as I return home : )